### PR TITLE
Add support for getting performance timing of filter callbacks

### PR DESCRIPTION
**Pull request by trevorforrey**
- Adds a new export `lfsDebug`, which when toggled will record performance of filter callbacks.

**Important**
This updates the shape that this library is imported, so we'll most likely want to set the base to a non-main base.

### CODE REVIEW

File: src/utils/spawnHelper.js [hunk 1]
```diff
@@ -51,6 +51,7 @@ const spawnCommand = (command, opts, stdin = '') => new Promise((resolve, reject
 });
 
 const spawn = async (command, stdin, opts = {}, credentialsCallback, repoPath = null) => {
+  console.log('spawning 2: ', command);
```
Comment (implausible) [inline src/utils/spawnHelper.js:54]: I think you left this in accidentally.

File: src/index.js [hunk 2]
```diff
@@ -50,37 +51,40 @@ LFS.prototype = {
...
-    setDefaultShellOptions(defaultShellOptions);
-  }
+    if (defaultShellOptions) {
+      setDefaultShellOptions(defaultShellOptions);
+    }
 
-  return _NodeGit;
+    return _NodeGit;
+  },
+  lfsDebug: LFS_DEBUG
```
Comment (implausible) [inline src/index.js:89]: Let's leave the original export alone, but attach `lfsDebug` to `module.exports` like this:

```javascript
module.exports = () => {} /* original function here */
module.exports.lfsDebug = LFS_DEBUG;
```

Remember, functions are first class objects in Javascript, so we can assign the debug property to the main export.

File: src/callbacks/apply.js [hunk 3]
```diff
@@ -46,17 +67,29 @@ export default (credentialsCallback) => {
...
+            const deltaTime = endTime - startTime;
+            LFS_DEBUG.recordEvent(event, {
+              source: source.path(),
+              duration: deltaTime
+            });
+          },
+          () => Error.CODE.OK,
+          () => Error.CODE.PASSTHROUGH
+        );
+    };
```
Comment (implausible) [inline src/callbacks/apply.js:92]: This doesn't look right. `.then` only takes 2 callbacks not 3 callbacks. If `finally` is available (check https://node.green/), do:
```javascript
.then(
  () => Error.Code.OK,
  () => Error.CODE.PASSTHROUGH
)
.finally(() => {/* debug goes here */});
```

**Review by implausible — CHANGES_REQUESTED** [review]

File: src/utils/spawnHelper.js (hunk 1 above)
Comment (trevorforrey) [reply src/utils/spawnHelper.js:54]: 🤦 . Thanks for catching this! 🙇

File: src/index.js (hunk 2 above)
Comment (trevorforrey) [reply src/index.js:89]: Great idea! Really cleans it up!

File: src/callbacks/apply.js (hunk 3 above)
Comment (trevorforrey) [reply src/callbacks/apply.js:92]: Checked `node.green`. Looks like the oldest version it is compatible with is `10.3.0`. Checked `nodegit` and we set `engines.node` to `">= 12.19.0 < 13 || >= 14.10.0"` so we should be good to go. Thanks for pointing this out. I had never seen this syntax before and thought this was `.then` taking a variable amount of functions to run but I should have looked up to confirm the syntax! Thanks again!

File: src/callbacks/apply.js [hunk 4]
```diff
@@ -7,6 +7,27 @@ import exec from '../utils/execHelper';
...
+    dumpEvents: (dumpPath) => fse.writeJsonSync(dumpPath, events, { spaces: 2 }),
+    recordEvent: (event, eventData) => {
+      if (shouldRecordEvents) {
+        events.push({
+          event,
+          data: eventData
+        });
+      }
+    },
+    retrieveEvents: () => [...events],
```
Comment (implausible) [inline src/callbacks/apply.js:24]: If this weren't a debug tool I would definitely have you adjust this, just fyi this is not equivalent to `fp.clone`. You are producing a new array reference, but each of the events here are referentially identical to the ones we have in the internal memory. This means that if you retrieveEvents and then modify any event in the list, you also modify the event in the internal memory.

This is definitely a big distinction, and you should look out for it on more critical code paths.

**Review by implausible — APPROVED** [review]

**Pull request merged**